### PR TITLE
[DNM][stdlib] Add extractAll(where:) method to RangeReplaceableCollection

### PR DESCRIPTION
**Pull request by Jumhyn**
This is the prototype implementation (and tests) for the [`extractAll(where:)` proposal](https://github.com/Jumhyn/swift-evolution/blob/extract-where/proposals/NNNN-extract-where-rangereplaceablecollection.md).

### CODE REVIEW

File: stdlib/public/core/RangeReplaceableCollection.swift [hunk 1]
```diff
@@ -1139,4 +1139,37 @@ extension RangeReplaceableCollection {
...
+  ///   sequence as its argument and returns a Boolean value indicating
+  ///   whether the element should be extracted from the collection.
+  ///
+  /// - Complexity: O(*n*), where *n* is the length of the collection.
+  @inlinable
+  public mutating func extractAll(
+    where shouldBeExtracted: (Element) throws -> Bool
+  ) rethrows -> Self {
+    var extracted = Self()
+    extracted.reserveCapacity(self.count)
```
Comment (Jumhyn) [inline stdlib/public/core/RangeReplaceableCollection.swift]: Is this too much overkill? Another option is to place all the extracted element into an array and then return `Self(array)` at the end. Not sure what the efficiency implications are there.

File: test/stdlib/ExtractElements.swift [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+extension RangeReplaceableCollection where Element: Equatable {
+  mutating func extract(equalTo: Element) -> Self {
+    return extractAll(where: { $0 == equalTo })
+  }
+}
+
+ExtractElements.test("extracting-array-with-predicate") {
+  var a = Array(0..<10)
+  let e = a.extractAll(where: { $0 % 2 == 0 })
+  expectEqualSequence([1,3,5,7,9], a)
```
Comment (Jumhyn) [inline test/stdlib/ExtractElements.swift]: This file mimics the tests in `RemoveElements.swift`. Should we combine the two?

File: test/stdlib/ExtractElements.swift [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+// RUN: %target-run-simple-swift
+// REQUIRES: executable_test
+
+import StdlibUnittest
+
+
+let ExtractElements = TestSuite("ExtractElements")
```
Comment (Jumhyn) [inline test/stdlib/ExtractElements.swift]: This file mimics the tests in `RemoveElements.swift`. Should we combine the two?

File: stdlib/public/core/RangeReplaceableCollection.swift [hunk 4]
```diff
@@ -872,6 +872,39 @@ extension RangeReplaceableCollection where Self : BidirectionalCollection {
...
+  /// - Parameter k: The number of elements to extract from the collection.
+  ///   `k` must be greater than or equal to zero and must not exceed the
+  ///   number of elements in the collection.
+  ///
+  /// - Complexity: O(*k*), where *k* is the specified number of elements.
+  @inlinable
+  public mutating func extractLast(_ k: Int) -> Self {
+    // Duplicate removeLast(_:) logic above to avoid redundant preconditions
+    // and index calculation
+    if k == 0 { return Self() }
```
Comment (theblixguy) [inline stdlib/public/core/RangeReplaceableCollection.swift]: A `guard` might be better/clearer: `guard x != 0 else { return Self() }`

File: stdlib/public/core/RangeReplaceableCollection.swift [hunk 5]
```diff
@@ -872,6 +872,39 @@ extension RangeReplaceableCollection where Self : BidirectionalCollection {
...
+    _precondition(k >= 0, "Number of elements to remove should be non-negative")
```
Comment (theblixguy) [inline stdlib/public/core/RangeReplaceableCollection.swift]: This should be `k > 0` as the check above takes care of the scenario where k == 0.

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

**Pull request closed without merging**